frontend/src/components/core/MainMenu/MainMenu.tsx

```diff
@@ -246,39 +223,61 @@ function MainMenu(props: Props): ReactElement {
     about: { onClick: props.aboutCallback, label: "About" },
   }
 
-  let menuOptions: any[] = [coreMenuOptions.rerun, coreMenuOptions.clearCache]
-
-  if (shouldShowS4AMenu) {
-    menuOptions = [
-      ...menuOptions,
-      coreMenuOptions.settings,
-      { ...coreMenuOptions.recordScreencast, hasDividerAbove: true },
-      ...S4AMenuOptions,
-    ]
-  } else {
-    menuOptions = [...menuOptions]
-
-    const showDeploy = props.isDeployable && isLocalhost()
-    if (showDeploy) {
-      menuOptions.push({ ...coreMenuOptions.deployApp, hasDividerAbove: true })
+  const S4AMenuOptions = props.s4aMenuItems.map(item => {
+    if (item.type === "separator") {
+      return coreMenuOptions.DIVIDER
+    }
+
+    return {
+      onClick: () =>
+        props.sendS4AMessage({
+          type: "MENU_ITEM_CALLBACK",
+          key: item.key,
+        }),
+      label: item.label,
     }
+  }, [] as any[])
+
+  const shouldShowS4AMenu = !!S4AMenuOptions.length
+
+  const preferredMenuOrder: any[] = [
```

Comment:
This looks so much easier to read, thank you!

One last minor thing to make it 100% clean, could we refactor out the boolean checks?

```
  const showDeploy = props.deployParams && isLocalhost() && !shouldShowS4AMenu
  const showSnapshot = props.sharingEnabled && !shouldShowS4AMenu
  const preferredMenuOrder: any[] = [
    coreMenuOptions.rerun,
    coreMenuOptions.clearCache,
    coreMenuOptions.DIVIDER,
    showDeploy && coreMenuOptions.deployApp,
    showSnapshot && coreMenuOptions.saveSnapshot,
    coreMenuOptions.recordScreencast,
...
```